### PR TITLE
Fix after review

### PR DESCRIPTION
**Pull request by mihaib79**

### CODE REVIEW

File: testApp/ledger/structs.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+// Write sends raw data to the device
+func (hf *hidFramer) Write(p []byte) (int, error) {
+	if debug {
+		fmt.Println("HID <=", hex.EncodeToString(p))
+	}
+	// split into 64-byte chunks
+	chunk := make([]byte, 64)
+	binary.BigEndian.PutUint16(chunk[:2], 0x0101)
```
Comment (andreibancioiu) [inline testApp/ledger/structs.go:55]: Code is complex. Can you think of a way to simplify it perhaps? Split into functions, avoid magic numbers etc.

File: testApp/ledger/structs.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+// Read reads raw data from the device
+func (hf *hidFramer) Read(p []byte) (int, error) {
+	if hf.seq > 0 && hf.pos != 64 {
+		// drain buf
+		n := copy(p, hf.buf[hf.pos:])
+		hf.pos += n
+		return n, nil
+	}
+	// read next 64-byte packet
```
Comment (andreibancioiu) [inline testApp/ledger/structs.go:80]: Just as you added these comments, perhaps it is possible to split the function into `readPacket`, `parseHeader`?

File: src/utils.c [hunk 3]
```diff
@@ -7,16 +7,43 @@
...
+uint32_t readUint32BE(uint8_t *buffer);
+void getAddressHexFromBinary(uint8_t *publicKey, char *address);
+void getAddressBech32FromBinary(uint8_t *publicKey, char *address);
+void getPublicKey(uint32_t accountNumber, uint32_t index, uint8_t *publicKeyArray);
+void getPrivateKey(uint32_t accountNumber, uint32_t index, cx_ecfp_private_key_t *privateKey);
+void sendResponse(uint8_t tx, bool approve);
+
+////////////////////////////////////////////////////////////////////////////////
+
+uint32_t readUint32BE(uint8_t *buffer) {
```
Comment (andreibancioiu) [inline src/utils.c:28]: Perhaps add comment?

File: testApp/main.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+	errGetAddress         = "Couldn't get address"
+	errGetBalanceAndNonce = "Couldn't get address balance and nonce"
+	errEmptyAddress       = "Empty address"
+	errInvalidAddress     = "Invalid receiver address"
+	errInvalidAmount      = "Invalid ERD amount"
+	errSigningTx          = "Signing error"
+	errSendingTx          = "Error sending tx"
+)
+
+type transactionType struct {
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply testApp/main.go:47]: Done.

File: testApp/main.go [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+	addr := uint32(addressBytes[len(addressBytes)-1])
+	shard := addr & maskHigh
+	if shard > noOfShards-1 {
+		shard = addr & maskLow
+	}
+	return shard
+}
+
+func getDeviceInfo(nanos *ledger.NanoS) (network, contractData, account, addressIndex uint8, ledgerVersion string) {
+	// retrieving various informations from Ledger
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply testApp/main.go:109]: Done.

File: testApp/main.go [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+	appVersion, err := nanos.GetVersion()
+	if err != nil {
+		log.Fatalln(errGetAppVersion, err)
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply testApp/main.go:112]: Done.

File: testApp/main.go [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+	reader := bufio.NewReader(os.Stdin)
+	// read destination address
+	fmt.Print("Enter destination address: ")
+	strReceiverAddress, _ := reader.ReadString('\n')
+	if strReceiverAddress == "" {
+		log.Fatalln(errEmptyAddress)
+	}
+	strReceiverAddress = strReceiverAddress[:len(strReceiverAddress)-1]
+	var receiverAddress []byte
+	if strings.HasPrefix(strReceiverAddress, "terd") || strings.HasPrefix(strReceiverAddress, "erd") {
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply testApp/main.go:143]: Done.

File: testApp/main.go [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+	}
+	if err != nil || len(receiverAddress) != 32 {
+		log.Fatalln(errInvalidAddress, err)
+	}
+	strReceiverShard := getAddressShard(hex.EncodeToString(receiverAddress), noOfShards)
+	fmt.Printf("Receiver shard: %v\n\r", strReceiverShard)
+	// read amount
+	fmt.Print("Amount of ERD to send: ")
+	strAmount, _ := reader.ReadString('\n')
+	if strings.HasSuffix(strAmount, "\n") {
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply testApp/main.go:158]: For some unknown reasons, \n is not always appended.

File: testApp/main.go [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+	if err != nil {
+		log.Fatalln(errInvalidAmount)
+	}
+	bigFloatAmount := big.NewFloat(0).SetFloat64(amount)
+	denomination, _ := big.NewFloat(0).SetString(strDenomination)
+	bigFloatAmount.Mul(bigFloatAmount, denomination)
+	bigIntAmount := new(big.Int)
+	bigFloatAmount.Int(bigIntAmount)
+	var data string
+	if contractData == 1 {
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply testApp/main.go:171]: In our implementation, if contract data is disabled, you are not allowed to send a transaction with a non-empty data field.

File: Makefile [hunk 10]
```diff
@@ -20,7 +20,7 @@ $(error Environment variable BOLOS_SDK is not set)
 endif
 include $(BOLOS_SDK)/Makefile.defines
 
-APP_LOAD_PARAMS= --curve ed25519 --path "44'/476'" --appFlags 0x240 $(COMMON_LOAD_PARAMS)
+APP_LOAD_PARAMS= --curve ed25519 --path "44'/508'" --appFlags 0x240 $(COMMON_LOAD_PARAMS)
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply Makefile:23]: All good :)

File: testApp/main.go [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+	fmt.Printf("Pubkey: %s\n\r", strPubkey)
+
+	// retrieve sender's nonce and balance
+	denomination, _ := big.NewFloat(0).SetString(strDenomination)
+	balance, nonce, err := getSenderInfo(strAddress)
+	if err != nil || balance == nil {
+		log.Fatalln(errGetBalanceAndNonce, err)
+	}
+	bigFloatBalance, _ := big.NewFloat(0).SetString(balance.String())
+	bigFloatBalance.Quo(bigFloatBalance, denomination)
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply testApp/main.go:247]: It looks prettier and it's easier for the user to input the correct amount of ERD he wants to send.

File: src/menu.c [hunk 12]
```diff
@@ -4,195 +4,200 @@
...
+UX_FLOW(ux_idle_flow,
+    &ux_idle_flow_1_step,
+    &ux_idle_flow_2_step,
+    &ux_idle_flow_3_step,
+    &ux_idle_flow_4_step,
+    FLOW_LOOP
+);
 
+//////////////////////////////////////////////////////////////////////////////////////
+// Enable contract data submenu:
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/menu.c:87]: Fixed.

File: src/utils.c (hunk 3 above)
Comment (mihaib79) [reply src/utils.c:28]: Done.

File: testApp/main.go [hunk 13]
```diff
@@ -0,0 +1,266 @@
+package main
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply testApp/main.go]: Done.

File: src/viewAddress.c [hunk 14]
```diff
@@ -5,35 +5,37 @@
 
 static char address[FULL_ADDRESS_LENGTH];
 
+void viewAddressAsBech32(uint32_t account, uint32_t index);
```
Comment (AdoAdoAdo) [inline src/viewAddress.c:8]: function declaration can be removed as it is included from viewAddress.h

File: src/viewAppVersion.c [hunk 15]
```diff
@@ -3,6 +3,10 @@
 #include "ux.h"
 #include "utils.h"
 
+static unsigned int app_version_ui_button(unsigned int button_mask, unsigned int button_mask_counter);
+void viewAppVersion();
```
Comment (AdoAdoAdo) [inline src/viewAppVersion.c:7]: function declaration can be removed as it is included from viewAppVersion.h

File: testApp/cmd/testApp/main.go [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+	gasPrice        = uint64(100000000000000)
+	gasLimit        = uint64(100000)
+	gasPerDataByte  = uint64(1500)
+)
+
+var network = [...]string{"Mainnet", "Testnet"}
+var status = [...]string{"Disabled", "Enabled"}
+
+const (
+	errOpenDevice         = "Couldn't open device"
```
Comment (AdoAdoAdo) [inline testApp/cmd/testApp/main.go:38]: error messages should start with lowercase

File: testApp/cmd/testApp/main.go [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+	"github.com/ElrondNetwork/ledger-elrond/testApp/ledger"
+
+	"github.com/btcsuite/btcutil/bech32"
+)
+
+const observerHost string = "https://wallet-api.elrond.com"
+const isProxy bool = true // if true, send "data" field as string, else send as base64
+
+const (
+	noOfShards      = uint32(5)
```
Comment (AdoAdoAdo) [inline testApp/cmd/testApp/main.go:27]: possibly different, maybe we should read this from the api.
@andreibancioiu was "https://wallet-api.elrond.com" the correct url?

File: testApp/cmd/testApp/main.go [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+	defer resp.Body.Close()
+	if err != nil {
+		return nil, 0, err
+	}
+	var balanceMsg accountMessage
+	err = json.Unmarshal(body, &balanceMsg)
+	if err != nil {
+		return nil, 0, err
+	}
+	balance, _ := big.NewInt(0).SetString(balanceMsg.Account.Balance, 10)
```
Comment (AdoAdoAdo) [inline testApp/cmd/testApp/main.go:93]: maybe should not ignore the error

File: testApp/cmd/testApp/main.go [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+
+	return balance, balanceMsg.Account.Nonce, nil
+}
+
+// getAddressShard returns the assigned shard of an address
+func getAddressShard(bech32Address string, noOfShards uint32) uint32 {
+	// convert sender from bech32 to hex pubkey
+	_, pubkeyBech32, _ := bech32.Decode(bech32Address)
```
Comment (AdoAdoAdo) [inline testApp/cmd/testApp/main.go:101]: maybe add error as return as well and treat errors inside

File: testApp/cmd/testApp/main.go [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+		if strings.HasSuffix(data, "\n") {
+			data = strings.TrimSuffix(data, "\n")
+		}
+	}
+	return strReceiverAddress, bigIntAmount, data, nil
+}
+
+// signTransaction sends the tx to Ledger for user confirmation and signing
+func signTransaction(tx *transaction, nanos *ledger.NanoS) error {
+	toSign, _ := json.Marshal(&tx)
```
Comment (AdoAdoAdo) [inline testApp/cmd/testApp/main.go:181]: tx is aleady a pointer, I don't think the address is required as parameter to json.Marshal
also treat error

File: testApp/ledger/nanos.go [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+	codeInvalidMessage       = 0x6e02
+	codeInvalidP1            = 0x6e03
+	codeMessageTooLong       = 0x6e04
+	codeReceiverTooLong      = 0x6e05
+	codeAmountTooLong        = 0x6e06
+	codeContractDataDisabled = 0x6e07
+)
+
+const (
+	errBadConfigResponse  = "GetConfiguration erroneous response"
```
Comment (AdoAdoAdo) [inline testApp/ledger/nanos.go:41]: lowercase for error string
is there a reason why errors.New is not used for these instead ?

File: src/getAddress.c [hunk 22]
```diff
@@ -5,60 +5,69 @@
...
 
-static uint8_t set_result_get_address() {
-    uint8_t tx = 0;
-    const uint8_t address_size = strlen(address);
-
-    G_io_apdu_buffer[tx++] = address_size;
-    os_memmove(G_io_apdu_buffer + tx, address, address_size);
-    tx += address_size;
+static uint8_t setResultGetAddress();
+void handleGetAddress(uint8_t p1, uint8_t p2, uint8_t *dataBuffer, uint16_t dataLength, volatile unsigned int *flags, volatile unsigned int *tx);
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/getAddress.c:9]: Done.

File: src/getAddress.c [hunk 23]
```diff
@@ -5,60 +5,69 @@
...
 
-    return tx;
-}
+////////////////////////////////////////////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/getAddress.c:11]: Done.

File: src/menu.c [hunk 24]
```diff
@@ -4,195 +4,196 @@
...
+const char* setting_network_getter(unsigned int idx);
+void setting_network_selector(unsigned int idx);
+void setting_contract_data_change(unsigned int enabled);
+const char* setting_contract_data_getter(unsigned int idx);
+void setting_contract_data_selector(unsigned int idx);
 const char* settings_submenu_getter(unsigned int idx);
 void settings_submenu_selector(unsigned int idx);
+const char* info_submenu_getter(unsigned int idx);
+void info_submenu_selector(unsigned int idx);
+void ui_idle(void);
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/menu.c:42]: Done.

File: src/menu.c [hunk 25]
```diff
@@ -4,195 +4,196 @@
...
+        break;
     case 1:
-      setting_network_change(1);
-      break;
+        setting_network_change(1);
+        break;
     default:
-      ux_menulist_init(0, settings_submenu_getter, settings_submenu_selector);
-  }
+        ux_menulist_init(0, settings_submenu_getter, settings_submenu_selector);
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/menu.c:111]: Done.

File: src/menu.c [hunk 26]
```diff
@@ -4,195 +4,196 @@
...
+        break;
     case 1:
-      setting_contract_data_change(1);
-      break;
+        setting_contract_data_change(1);
+        break;
     default:
-      ux_menulist_init(0, settings_submenu_getter, settings_submenu_selector);
-  }
+        ux_menulist_init(0, settings_submenu_getter, settings_submenu_selector);
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/menu.c:141]: Done.

File: src/menu.c [hunk 27]
```diff
@@ -4,195 +4,196 @@
...
+        break;
     case 2:
-      ux_menulist_init_select(0, setting_contract_data_getter, setting_contract_data_selector, N_storage.setting_contract_data);
-      break;
+        ux_menulist_init_select(0, setting_contract_data_getter, setting_contract_data_selector, N_storage.setting_contract_data);
+        break;
     default:
-      ui_idle();
-  }
+        ui_idle();
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/menu.c:165]: Done.

File: testApp/cmd/testApp/main.go (hunk 18 above)
Comment (mihaib79) [reply testApp/cmd/testApp/main.go:93]: Done.

File: testApp/cmd/testApp/main.go (hunk 19 above)
Comment (mihaib79) [reply testApp/cmd/testApp/main.go:101]: Done.

File: testApp/cmd/testApp/main.go (hunk 20 above)
Comment (mihaib79) [reply testApp/cmd/testApp/main.go:181]: Done.

File: src/menu.c [hunk 28]
```diff
@@ -4,195 +4,196 @@
...
-      break;
-    case 2:
-      viewAppVersion();
-      break;
+        viewAppVersion();
+        break;
     default:
-      ui_idle();
-  }
+        ui_idle();
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/menu.c:175]: Done.

File: src/selectAccount.c [hunk 29]
```diff
@@ -3,23 +3,20 @@
...
+
+bool menu_selection;
 static char account[4], address_index[4]; // hold the int_to_string representations
 char intAccount, intAddressIndex; // temporary copies of NV variables
 
+void intToString();
+static unsigned int select_address_index_ui_button(unsigned int button_mask, unsigned int button_mask_counter);
+static unsigned int select_account_ui_button(unsigned int button_mask, unsigned int button_mask_counter);
 void ui_address_menu_idle();
+void selectAccount();
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/selectAccount.c]: Done.

File: src/signTx.h [hunk 30]
```diff
@@ -2,16 +2,21 @@
...
 
+#ifndef _SIGN_TX_H_
+#define _SIGN_TX_H_
+
 void handleSignTx(uint8_t p1, uint8_t p2, uint8_t *dataBuffer, uint16_t dataLength, volatile unsigned int *flags, volatile unsigned int *tx);
 
 #define MAX_AMOUNT_LEN 32
 #define MAX_BUFFER_LEN 1024
 
 typedef struct {
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/signTx.h]: Done.

File: src/signTx.c [hunk 31]
```diff
@@ -6,15 +6,15 @@
...
-    tx += sig_size;
-    return tx;
-}
+static uint8_t setResultSignature();
+void makeAmountPretty();
+static int jsoneq(const char *json, jsmntok_t *tok, const char *s);
+uint16_t txDataReceived(uint8_t *dataBuffer, uint16_t dataLength);
+uint16_t parseData();
+void signTx();
+void handleSignTx(uint8_t p1, uint8_t p2, uint8_t *dataBuffer, uint16_t dataLength, volatile unsigned int *flags, volatile unsigned int *tx);
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/signTx.c]: Done.

File: src/signTx.c [hunk 32]
```diff
@@ -173,6 +170,38 @@ void handleSignTx(uint8_t p1, uint8_t p2, uint8_t *dataBuffer, uint16_t dataLeng
...
+        tx_context.bufLen = 0;
+        retCode = txDataReceived(dataBuffer, dataLength);
+        break;
+    // consequent packets if the tx is larger than 255 bytes
+    case P1_MORE:
+        retCode = txDataReceived(dataBuffer, dataLength);
+        break;
+    default:
+        // unknown packet type
+        retCode = ERR_INVALID_P1;
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/signTx.c:190]: Done.

File: src/utils.c [hunk 33]
```diff
@@ -7,16 +7,43 @@
...
+  0 | HARDENED_OFFSET,
+  0 | HARDENED_OFFSET
+};
+
+uint32_t readUint32BE(uint8_t *buffer);
+void getAddressHexFromBinary(uint8_t *publicKey, char *address);
+void getAddressBech32FromBinary(uint8_t *publicKey, char *address);
+void getPublicKey(uint32_t accountNumber, uint32_t index, uint8_t *publicKeyArray);
+void getPrivateKey(uint32_t accountNumber, uint32_t index, cx_ecfp_private_key_t *privateKey);
+void sendResponse(uint8_t tx, bool approve);
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/utils.c]: Done.

File: src/viewAddress.c (hunk 14 above)
Comment (mihaib79) [reply src/viewAddress.c]: Done.

File: src/viewAppVersion.c (hunk 15 above)
Comment (mihaib79) [reply src/viewAppVersion.c]: Done.

File: testApp/cmd/testApp/main.go (hunk 16 above)
Comment (mihaib79) [reply testApp/cmd/testApp/main.go]: Done.

File: testApp/ledger/nanos.go (hunk 21 above)
Comment (mihaib79) [reply testApp/ledger/nanos.go:41]: Solved.

File: testApp/cmd/testApp/main.go [hunk 34]
```diff
@@ -0,0 +1,308 @@
...
+		return err
+	}
+	fmt.Println("Signing transaction. Please confirm on your Ledger")
+	signature, err := nanos.SignTxn(toSign)
+	if err != nil {
+		log.Println(errSigningTx)
+		return err
+	}
+
+	// send transaction
```
Comment (AdoAdoAdo) [inline testApp/cmd/testApp/main.go:212]: update comment

File: testApp/cmd/testApp/main.go (hunk 17 above)
Comment (mihaib79) [reply testApp/cmd/testApp/main.go:27]: Will fix this after the route with this information will be added to the proxy.

File: testApp/cmd/testApp/main.go (hunk 34 above)
Comment (mihaib79) [reply testApp/cmd/testApp/main.go:212]: Done.

File: src/menu.c [hunk 35]
```diff
@@ -4,195 +4,196 @@
...
+        return setting_network_getter_values[idx];
+    return NULL;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+
 void setting_network_selector(unsigned int idx) {
-  switch(idx) {
+    switch(idx) {
     case 0:
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/menu.c:97]: Done.

File: src/menu.c [hunk 36]
```diff
@@ -4,195 +4,196 @@
...
+        return setting_contract_data_getter_values[idx];
+    return NULL;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+
 void setting_contract_data_selector(unsigned int idx) {
-  switch(idx) {
+    switch(idx) {
     case 0:
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply src/menu.c:123]: Done.

**Pull request merged**